cparser/Elab.ml

```diff
@@ -1793,45 +1853,32 @@ let elab_expr vararg loc env a =
 
 (* Elaboration of pre- or post- increment/decrement *)
   and elab_pre_post_incr_decr op msg a1 =
-      let b1,env = elab env a1 in
-      if not (is_modifiable_lvalue env b1) then
-        err "the argument of %s is not a modifiable l-value" msg;
-      if not (is_scalar_type env b1.etyp) then
-        err "the argument of %s must be an arithmetic or pointer type" msg;
-      { edesc = EUnop(op, b1); etyp = b1.etyp },env
+    let b1,env = elab env a1 in
+    if not (is_modifiable_lvalue env b1) then
+      err "expression is not assignable";
+    if not (is_scalar_type env b1.etyp) then
+      err "cannot %s value of type %a" msg (print_typ env) b1.etyp;
+    { edesc = EUnop(op, b1); etyp = b1.etyp },env
 
 (* Elaboration of binary operators over integers *)
-  and elab_binary_integer msg op a1 a2 =
-      let b1,env = elab env a1 in
-      if not (is_integer_type env b1.etyp) then
-        error "the first argument of '%s' is not an integer type" msg;
-      let b2,env = elab  env a2 in
-      if not (is_integer_type env b2.etyp) then
-        error "the second argument of '%s' is not an integer type" msg;
-      let tyres = binary_conversion env b1.etyp b2.etyp in
-      { edesc = EBinop(op, b1, b2, tyres); etyp = tyres },env
+  and elab_binary_integer op a1 a2 =
+    let b1,env = elab env a1 in
+    let b2,env = elab env a2 in
+    if not ((is_integer_type env b1.etyp) && (is_integer_type env b2.etyp)) then
+      error "invalid operands to binary expression (%a and %a)"
+        (print_typ env) b1.etyp (print_typ env) b2.etyp;
+    let tyres = binary_conversion env b1.etyp b2.etyp in
+    { edesc = EBinop(op, b1, b2, tyres); etyp = tyres },env
 
 (* Elaboration of binary operators over arithmetic types *)
-  and elab_binary_arithmetic msg op a1 a2 =
-      let b1,env = elab env a1 in
-      if not (is_arith_type env b1.etyp) then
-        error "the first argument of '%s' is not an arithmetic type" msg;
-      let b2,env = elab env a2 in
-      if not (is_arith_type env b2.etyp) then
-        error "the second argument of '%s' is not an arithmetic type" msg;
-      let tyres = binary_conversion env b1.etyp b2.etyp in
-      { edesc = EBinop(op, b1, b2, tyres); etyp = tyres },env
-
-(* Elaboration of shift operators *)
-  and elab_shift msg op a1 a2 =
-      let b1,env = elab env a1 in
-      if not (is_integer_type env b1.etyp) then
-        error "the first argument of '%s' is not an integer type" msg;
-      let b2,env = elab env a2 in
-      if not (is_integer_type env b2.etyp) then
-        error "the second argument of '%s' is not an integer type" msg;
-      let tyres = unary_conversion env b1.etyp in
```

Comment:
Note: this is a unary conversion of the first argument's type.  That's what makes shifts different from other binary arithmetic operators.